src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -90,7 +94,13 @@ internal partial class CurlHandler
             try
             {
                 // Set task completion after all headers have been received
-                // TODO: Fail if we find that task is already Canceled or Faulted
+                // Fail if we find that task is already Canceled or Faulted
+                if (state.Task.IsCanceled || state.Task.IsFaulted)
+                {
+                    // Returing a value other than size fails the callback and forces
+                    // request completion with an error
+                    return (size > 0) ? size - 1 : 1;
+                }
```

Comment:
Is there a race condition here where the task could be completed as canceled/faulted after the checks, or is it guaranteed that it would have already completed that way if it was going to?  If there is a race condition, you could rewrite this as the following to avoid it:
```C#
if (!state.TrySetResult(state.ResponseMessage) &&
    (state.Task.IsCanceled || state.Task.IsFaulted))
{
    return size > 0 ? size - 1 : 1;
}
```